### ISSUE TITLE
Improve `http` resource

### ISSUE THREAD

**Issue by chris-rock** [inspec/inspec#1649]
### Description

Right now, we cover basic tests with `http` only:

```
describe http('http://localhost:8080/ping', auth: {user: 'user', pass: 'test'}, params: {format: 'html'}) do
  its('status') { should cmp 200 }
  its('body') { should cmp 'pong' }
  its('headers.Content-Type') { should cmp 'text/html' }
end

describe http('http://example.com/ping').headers do
  its('Content-Length') { should cmp 258 }
  its('Content-Type') { should cmp 'text/html; charset=UTF-8' }
end
```

It would be nice if we could make testing via `http` easier.

### InSpec and Platform Version

1.19.2

### Possible Solutions

**Write a resource around faraday and reuse its interface**

To highlight the reuse of faraday, I am going to use `faraday` as the resource name instead of `http`

```
describe faraday(:url => 'http://sushi.com').get '/nigiri/sake.json' do
  its('body') { should cmp 'pong' }
end      
```

```
describe faraday(:url => 'http://sushi.com').get '/nigiri', { name: 'Maguro' }
  its('status') { should cmp 'pong' }
  its('body') { should cmp 'pong' }
end
```

```
describe faraday(:url => 'http://sushi.com').post {url: '/nigiri', headers: { 'Content-Type' => 'application/json' }, body: '{ "name": "Unagi" }' }
  its('body.name') { should eq "Unagi" }
end
```

**Support for custom ca_path**

Sometimes we have custom `ca_path`. See https://github.com/lostisland/faraday/wiki/Setting-up-SSL-certificates for full description. We would follow the faraday pattern too:

```
cert_path = "~/.chef/trusted_certs"

describe faraday(url: 'https://encrypted.google.com', ssl: { ca_path: "/usr/lib/ssl/certs" }).get '/nigiri/sake.json' do
  its('body') { should cmp 'pong' }
end   
```

**Comment by chris-rock** [comment]
Just to be clear. Although I used `faraday` as the resource name, this was just used to highlight our reuse of faraday gem. The resource name will stay `http`

**Comment by wswheeler** [comment]
I started using the http resource as part of running tests against webservers that I was building up with Chef. The feature that I think would be most useful is if there was a way to lie about DNS to validate SSL certificates as well as well configurations. For example, I'd like to go to the full host name with https and validate that it connected and additionally got the expected content. Right now I'm faking this with the following type of describe. 

```
  describe http('https://127.0.0.1/',
                headers: {'Host' => 'web.example.com'},
                ssl_verify: false) do
    its('status') { should eq 200 }
    its('body') { should eq 'web.example.com content' }
  end
```
I would like to be able to do something more like
```
  describe http('https://web.example.com/', resolve_hostname_to: '127.0.0.1' ) do
    its('status') { should eq 200 }
    its('body') { should eq 'web.example.com content' }
  end
```
Or maybe even:
```
  describe http('https://web.example.com/', use_localhost: true ) do
    its('status') { should eq 200 }
    its('body') { should eq 'web.example.com content' }
  end
```
Either of the two examples would allow SSL verification to be true and would set the Host header to the value provided as part of the URL. Currently I know that you can do this with the curl tool as the following:
```
curl --resolve web.example.com:443:127.0.0.1 https://web.example.com
```

**Comment by kekaichinose** [comment]
Adding some additional color on this one, but minus the ask from @wswheeler being addressed in #4128, this issue has aged beyond the version for which it was cut against.

__What’s happening? Why was this issue closed?__
This issue was closed due to some much needed review of legacy issues or issues that were spawned in older versions of InSpec, i.e. < v3.

__Why do I care?__
You would care about this if this was an issue that you are still seeing and/or feel needs to be addressed in the current version of InSpec.

__What do I need to do?__
If this issue is no longer important, no further action is necessary. However, if you think this is something that should be addressed, please reopen the issue and continue the discussion.

Thank you!
__Chef InSpec Team__
